### PR TITLE
build: Update api gomod dependencies

### LINKED ISSUE #15102: Update gomod dependencies to latest before 1.16 release
dependabot seems to be stuck updating go deps. It would be a good practice to make sure these are updated to latest before the 1.16 release

https://github.com/rook/rook/pull/15052
https://github.com/rook/rook/pull/15081

@obnoxxx is this a release task you could help out with?

### PR DESCRIPTION
**Pull request by obnoxxx**
**Description of changes:** 

This updates the go module dependencies under `pkg/apis`.

**Issue resolved by this Pull Request:**
Resolves #15102 

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

**Review by BlaineEXE — CHANGES_REQUESTED** [review]

File: pkg/apis/go.mod [hunk 1]
```diff
@@ -1,8 +1,8 @@
 module github.com/rook/rook/pkg/apis
 
-go 1.22.5
+go 1.23
 
-toolchain go1.22.7
+toolchain go1.23.1
```
[the remark this answers is not in the document]
Comment (obnoxxx) [reply pkg/apis/go.mod]: done. reverted.

[the remark this answers is not in the document]
Comment (travisn) [reply pkg/apis/go.mod]: A few weeks back discussing with Subham, it seemed to update some of these go dependencies we have to move to 1.23 and cut ties with 1.22. We were blocked from that downstream so we were holding off. I pinged someone today about that to see if that's still the case. But for upstream I think we're not blocked from this change and I'd suggest we're at the point where we can go ahead and restrict just to go 1.23 for the 1.16 release.

**Comment by BlaineEXE** [comment]
@travisn how do you feel about the `build(deps)` commit subject? Commitlint accepts it, but we don't really use that from what I remember.

**Comment by BlaineEXE** [comment]
@obnoxxx I poked around and found out the fix. This is a really strange error to come up (below), and it doesn't do a good job of saying what is wrong exactly. But the fix is easy: update the go version in all the github actions from 1.22 to 1.23. Of note, be sure all `1.22` and `1.22.*` strings are updated to `"1.23"`. There's currently a `1.22.z` input used, but we do want to change that to just `"1.23"` with no Z version set.

```
Generating deepcopy code for 1 targets
2024/12/09 19:00:01 internal error: package "sigs.k8s.io/json" without types was imported from "k8s.io/apimachinery/pkg/util/json"
```

**Comment by travisn** [comment]
> @travisn how do you feel about the `build(deps)` commit subject? Commitlint accepts it, but we don't really use that from what I remember.

Agreed, let's just use the `build` prefix and leave the other pattern to dependabot

**Comment by obnoxxx** [comment]
@BlaineEXE  qrote:
> > @travisn how do you feel about the `build(deps)` commit subject? Commitlint accepts it, but we don't really use that from what I remember.
> 
> Agreed, let's just use the `build` prefix and leave the other pattern to dependabot


done

**Comment by obnoxxx** [comment]
@BlaineEXE  wrote:
> @obnoxxx I poked around and found out the fix. This is a really strange error to come up (below), and it doesn't do a good job of saying what is wrong exactly. But the fix is easy: update the go version in all the github actions from 1.22 to 1.23. Of note, be sure all `1.22` and `1.22.*` strings are updated to `"1.23"`. There's currently a `1.22.z` input used, but we do want to change that to just `"1.23"` with no Z version set.

thanks for the analysis! I added the update to 1.23 as a separate commit. Let me know if you'd like me to squash



> 
> ```
> Generating deepcopy code for 1 targets
> 2024/12/09 19:00:01 internal error: package "sigs.k8s.io/json" without types was imported from "k8s.io/apimachinery/pkg/util/json"
> ```

**Comment by obnoxxx** [comment]
surprisingly, unit tests fail in the ci. more precisely:
`FAIL: TestUpdateCsiClusterConfig/test_empty_namespace_correction`

more detail:
```console
        	Error Trace:	/home/runner/work/rook/rook/pkg/operator/ceph/csi/cluster_config_test.go:457
        	Error:      	Not equal: 
```

**Comment by obnoxxx** [comment]
unit tests fail the exact same way locally for me on current master if I run `make mod.update` before `make test`

**Comment by BlaineEXE** [comment]
It should be the same commit because the go min version is being updated to 1.23. It's a required consequence, so it shouldn't be a separate commit.

**Comment by obnoxxx** [comment]
@BlaineEXE  wrote:

> It should be the same commit because the go min version is being updated to 1.23. It's a required consequence, so it shouldn't be a separate commit.

Okay, squashed.

**Comment by obnoxxx** [comment]
rebased after #15132  got merged.

File: go.mod [hunk 2]
```diff
@@ -79,19 +77,18 @@ require (
...
 require (
 	emperror.dev/errors v0.8.1 // indirect
-	github.com/Azure/go-ansiterm v0.0.0-20210617225240-d185dfc1b5a1 // indirect
+	github.com/Azure/go-ansiterm v0.0.0-20230124172434-306776ec8161 // indirect
 	github.com/ansel1/merry v1.8.0 // indirect
-	github.com/ansel1/merry/v2 v2.2.0 // indirect
+	github.com/ansel1/merry/v2 v2.2.1 // indirect
 	github.com/beorn7/perks v1.0.1 // indirect
-	github.com/ceph/ceph-csi/api v0.0.0-20231227104434-06f9a98b7a83
+	github.com/ceph/ceph-csi/api v0.0.0-20241209162139-8ffd9f515a28
```
Comment (BlaineEXE) [inline go.mod:84]: @travisn are we still planning to leave this dependency at its old version and have the CSI team do this update?

Comment (travisn) [reply go.mod:84]: Discussion in #15128 is that we already covered everything csi recommended, so we will just keep it in this PR.

**Review by travisn — APPROVED** [review]
LGTM!

File: pkg/apis/go.mod (hunk 1 above)
[the remark this answers is not in the document]
Comment (obnoxxx) [reply pkg/apis/go.mod]: @travisn , @BlaineEXE  I just added the result of `make mod.update`back into ythe commit which re-adds the toolchain go version update.

**Review by travisn — CHANGES_REQUESTED** [review]
Some CI issues to resolve now

**Comment by obnoxxx** [comment]
@travisn  wrote:
> Some CI issues to resolve now

addressed.

**Review by travisn — APPROVED** [review]
CI is green again

**Pull request merged**